presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1626,25 +1654,24 @@ private void analyzeAggregations(QuerySpecification node,
             }
         }
 
-        ImmutableList<FieldOrExpression> allGroupingColumns = groupingSets.stream()
-                .flatMap(Collection::stream)
-                .distinct()
-                .collect(toImmutableList());
-
         // is this an aggregation query?
-        if (!aggregates.isEmpty() || !allGroupingColumns.isEmpty()) {
+        if (!aggregates.isEmpty() || !(groupingSets.size() == 1 && getOnlyElement(groupingSets).isEmpty())) {
```

Comment:
The intention of this line is very non-obvious. Here is what it means:
* `!aggregates.isEmpty` checks implicit aggregation: query shape `SELECT sum(x) FROM t`.
* `!allGroupingColumns.isEmpty()` checks explicit aggregation: query shape `SELECT ... FROM t GROUP BY ...`

What you changed it to is not what this line intends to do.

In fact, this line was wrong before this pull request. They line incorrectly treats `group by ()`, which was not a valid syntax before you introduced it. At the time, I missed this during the code review.

Please fix the bug, preferably in a separate commit/pr. That should only affect a few lines. So it should be trivial to review. In that fix, also document what this line of code is doing. This line definitely needs comment.